src/generic_event_type.cpp

```diff
@@ -0,0 +1,374 @@
+bool string_id<generic_event_type>::is_valid() const
+{
+    return generic_event_type_factory.is_valid( *this );
+}
+
+void generic_event_type::finalize()
+{
+
+}
+
+void generic_event_type::check() const
+{
+    for( const effect_info &info : effects_to_add ) {
+        if( !info.id.is_valid() ) {
+            debugmsg( "Effect %s does not exist.", info.id.c_str() );
+            abort();
+        }
+    }
+    for( const efftype_id &id : effects_to_remove ) {
+        if( !id.is_valid() ) {
+            debugmsg( "Effect %s does not exist.", id.c_str() );
+            abort();
+        }
+    }
+    for( const trait_id &id : traits_to_remove ) {
+        if( !id.is_valid() ) {
+            debugmsg( "trait %s does not exist.", id.c_str() );
+            abort();
+        }
+    }
+    for( const trait_id &id : traits_to_add ) {
+        if( !id.is_valid() ) {
+            debugmsg( "trait %s does not exist.", id.c_str() );
+            abort();
+        }
+    }
+    for( const bionic_id &id : cbms_to_add ) {
+        if( !id.is_valid() ) {
+            debugmsg( "cbm %s does not exist.", id.c_str() );
+            abort();
+        }
+    }
+    for( const bionic_id &id : cbms_to_remove ) {
+        if( !id.is_valid() ) {
+            debugmsg( "cbm %s does not exist.", id.c_str() );
+            abort();
+        }
+    }
+    if( !weather_change.is_empty() && !weather_change.is_valid() ) {
+        debugmsg( "weather type %s does not exist.", weather_change.c_str() );
+        abort();
+    }
+    if( !target_part.is_empty() && !target_part.is_valid() ) {
+        debugmsg( "Target part %s does not exist.", target_part.c_str() );
+        abort();
+    }
+    for( const spawn_type &spawn : spawns ) {
+        if( !spawn.target.is_empty() && !spawn.target.is_valid() ) {
+            debugmsg( "Spawn target %s does not exist.", spawn.target.c_str() );
+            abort();
+        }
+    }
+    for( const generic_event_type_field &field : fields ) {
+        if( !field.type.is_valid() ) {
+            debugmsg( "field type %s does not exist.", field.type.c_str() );
+            abort();
+        }
+    }
+}
+
+void generic_event_type::load( const JsonObject &jo, const std::string & )
+{
+    optional( jo, was_loaded, "message", message );
+    optional( jo, was_loaded, "sound_message", sound_message );
+    optional( jo, was_loaded, "sound_effect", sound_effect, "" );
+    optional( jo, was_loaded, "lightning", lightning, false );
+    optional( jo, was_loaded, "pain", pain, 0 );
+    optional( jo, was_loaded, "wet", wet, 0 );
+    optional( jo, was_loaded, "radiation", radiation, 0 );
+    optional( jo, was_loaded, "healthy", healthy, 0 );
+    optional( jo, was_loaded, "weather_change", weather_change, WEATHER_NULL );
+    optional( jo, was_loaded, "update_weather", update_weather, false );
+    for( const std::string &trait : jo.get_string_array( "traits_to_add" ) ) {
+        traits_to_add.push_back( trait_id( trait ) );
+    }
```

Comment:
```suggestion
    optional( jo, was_loaded, "traits_to_add", traits_to_add );
```
You can load json arrays directly into C++ containers; you shouldn't need all these `for` loops in this function.